events/events.go

```diff
@@ -31,9 +35,83 @@ type EventListener interface {
 	HandleEvent(event Event)
 }
 
+// EventEmitter describes an interface that can be used to emit events
+type EventEmitter interface {
+	EmitEvent(Event)
+}
+
 // EventRegistrar is an object that you can register EventListeners on.
 type EventRegistrar interface {
 	RegisterListener(EventListener)
+	DeregisterListener(EventListener)
+}
+
+type baseEventRegistrar struct {
+	lock      sync.RWMutex
+	listeners []EventListener
+}
+
+// RegisterListener adds a listener to the Event Registar. Events emitted on this
+// registar will be invoked on the listener
+func (a *baseEventRegistrar) RegisterListener(l EventListener) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// by making a copy the backing array will never be changed after its creation.
+	// this allowes to copy the slice while locked but iterate while not locked
+	// preventing deadlocks when listeners are added/removed in the handler of a
+	// listener
+	cpy := append([]EventListener(nil), a.listeners...)
+	cpy = append(cpy, l)
+	a.listeners = cpy
+}
+
+// DeregisterListener removes a listener from the Event Registar. New events
+// emitted to this registar will not be invoked on the listener.
```

Comment:
This comment is not guaranteed to be true because you're copying the array. If an event is emitted right before a listener is removed, it's possible that the event is still gonna be sent to the listener.
Maybe clarify it by saying something like "events emitted after removal will not be invoked on the listener anymore. note: events that are already emitted but are not yet invoked on the listener might be invoked even though the listener is already removed."